### ISSUE TITLE
read_csv reads header row multiple times

### ISSUE THREAD

**Issue by mfaafm** [dask/dask#2976]
When reading a csv file with the `read_csv` method of `dask.dataframe` the header row is read multiple times. This happens when the `names` argument is set but `header` is not set to `None`.

Versions:
* dask==0.16.0
* pandas==0.21.0

**Example:**

```python
import dask.dataframe as dd
import pandas as pd

# create csv file
filename = 'test.csv'
n_rows = 1000

first_line = "1,1.0,3\n"
content = "4,5.0,6\n"
with open('test.csv', 'w') as f:
    f.write(first_line)
    f.write(content*(n_rows - 1))

names = ['A', 'B', 'C']
dtypes = {'A': 'uint32', 'B': 'float64', 'C': 'uint32'}

df_pandas = pd.read_csv(filename, sep=',', names=names, dtype=dtypes)
count_pandas = df_pandas.A.count()
groupby_pandas = df_pandas.groupby('B').count()

df_dask = dd.read_csv(filename, sep=',', names=names, dtype=dtypes, blocksize=100)
count_dask = df_dask.A.count().compute()
groupby_dask = df_dask.groupby('B').count().compute()

df_dask_correct = dd.read_csv(filename, sep=',', header=None, names=names, dtype=dtypes, blocksize=100)
count_dask_correct = df_dask_correct.A.count().compute()
groupby_dask_correct = df_dask_correct.groupby('B').count().compute()
```

The correct output, which is the same for `count_dask_correct` and `groupby_dask_correct`, is:
```python
print(count_pandas)
print(groupby_pandas)
```
1000

<table>
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>A</th>
      <th>C</th>
    </tr>
    <tr>
      <th>B</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1.0</th>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>5.0</th>
      <td>999</td>
      <td>999</td>
    </tr>
  </tbody>
</table>

However, we get:
```python
print(count_dask)
print(groupby_dask)
```
    1079

<table>
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>A</th>
      <th>C</th>
    </tr>
    <tr>
      <th>B</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1.0</th>
      <td>80</td>
      <td>80</td>
    </tr>
    <tr>
      <th>5.0</th>
      <td>999</td>
      <td>999</td>
    </tr>
  </tbody>
</table>

Note, that the number of partitions is 80:
```python
print(df_dask.npartitions)
```
    80

**Comment by wehlutyk** [comment]
Just bumped into this, and I have another maybe simpler example.

When using the following `test.csv` file:
```csv
1,a
2,b
3,c
4,d
5,e
```

Reading the file into several partitions (hence the small blocksize) adds the first line to every partition:
```python
In [1]: import dask.dataframe as dd
   ...: df = dd.read_csv(['test.csv'], names=['A', 'B'], blocksize=10)
   ...:
   ...: for i in range(df.npartitions):
   ...:     print('Partition {}'.format(i))
   ...:     print(df.get_partition(i).compute())
   ...:     print()
   ...:
Partition 0
   A  B
0  1  a
1  2  b
2  3  c

Partition 1
   A  B
0  1  a
1  4  d
2  5  e
```

The correct output should be
```
Partition 0
   A  B
0  1  a
1  2  b
2  3  c

Partition 1
   A  B
0  4  d
1  5  e
```

Explicitly setting `header=None` to `dd.read_csv()` solves the problem, but it's still unexpected behaviour.

**Comment by mrocklin** [comment]
I agree that this is an issue.  Thank you both for the reproducible bug reports.  Does anyone have the time and interest to look into a fix?

**Comment by jcrist** [comment]
Looks like we just need to handle `header='infer'` properly. Something like:

```python
if header == 'infer' and names is not None:
    header = None
```

earlier on in `read_pandas` should fix this.

**Comment by wehlutyk** [comment]
At least for the behaviour I saw, yes that should be enough! I don't know the code base at all, but is the fact that the first line was otherwise repeated in all partitions explained by the lack of that check? (i.e. is there not an underlying bug?)

**Comment by mfaafm** [comment]
Thanks for looking into this!

Header is set to `'infer'` by default, even if `names` is given. I think the issue is caused here:
```python
header = kwargs.get('header', 'infer')
need = 1 if header is None else 2
```
The solution of @jcrist should work. Alternatively, i think it could be fixed directly when reading the kwargs:
```python
header = kwargs.get('header', 'infer' if names is None else None)
```

**Comment by jcrist** [comment]
Fixed by #3022. Closing.
